### ISSUE TITLE
Enhancement: limit use of Centos 7 -D flags

### ISSUE THREAD

**Issue by lcn2** [ioccc-src/mkiocccentry#311]
### Background

With commit f85ec5ac7673281b9b2035de752f6df9be207902 we uncovered a significant problem with how C files are compiled while trying to also support old platforms such as **Centos 7**:

There is a significant problem with using:

```sh
    -D_DEFAULT_SOURCE -D_ISOC99_SOURCE -D_POSIX_C_SOURCE=200809L -D_XOPEN_SOURCE=600
```

These defines prevent functions that were introduced into C post c99
from being compiled without a warning/error.  Take for example, the
use of timegm(3) in entry_util.c:

```sh
    cc -D_DEFAULT_SOURCE -D_ISOC99_SOURCE -D_POSIX_C_SOURCE=200809L -D_XOPEN_SOURCE=600 -std=gnu11 -O0 -g -pedantic -Wall -Wextra -Werror -Wno-unused-command-line-argument -Wno-poison-system-directories -Wno-unreachable-code-break -Wno-padded -lm entry_util.c -c
```

will produce the following error:

```
    entry_util.c:1150:25: error: implicit declaration of function 'timegm' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        timestr_as_time_t = timegm(&timeptr);
```

Whereas this:

```sh
    cc -std=gnu11 -O0 -g -pedantic -Wall -Wextra -Werror -Wno-unused-command-line-argument -Wno-poison-system-directories -Wno-unreachable-code-break -Wno-padded -lm entry_util.c -c
```

will compile successfully.

While we understand that certain older systems such as CentOS 7 need:

```sh
    -D_DEFAULT_SOURCE -D_ISOC99_SOURCE -D_POSIX_C_SOURCE=200809L -D_XOPEN_SOURCE=600
```

some of the time, other systems **CANNOT** use them as they force the compiler,
in certain cases, to process system header files as of `-std=c99` were in effect
even though `-std=gnu11` is being used.

Even where modern systems do not generate warnibngs/errors, the use
of those defines could have adverse impacts in system include files
that might be incompatible with the C standard library being invoked
by the `-std `flag.

### Makefile changes

To address this somewhat, the following changes to `Makefile` were made:

The `${STD_SRC}` has been changed to:

```make
D_LEGACY= -D_DEFAULT_SOURCE -D_ISOC99_SOURCE -D_POSIX_C_SOURCE=200809L -D_XOPEN_SOURCE=600
C_STD= -std=gnu11
```

The `${CFLAGS}` is now:

```make
CFLAGS= ${C_STD} ${COPT} -pedantic ${WARN_FLAGS} ${LDFLAGS}
```

A new `${LEGACY_CFLAGS}` was added:

```make
LEGACY_CFLAGS= ${D_LEGACY} ${CFLAGS
```

To compile this change, all compile lines with `${CC}` use `${LEGACY_CFLAGS}`
except for entry_util.o which uses `${CFLAGS}`.

### Task needed

On a **Centos 7** platform, convert as many uses of `${LEGACY_CFLAGS}`
in the Makefile to just `${LEGACY_CFLAGS}` and still compile the repo.

Of if this is **NOT** possible, say because `entry_util.c` cannot be compiled
on **Centos 7**  with just `${CFLAGS}`, then choose to abandon the
**Centos 7** platform as a supported platform for this repo.

NOTE: If **Centos 7** were to be abandoned, **people could still submit**
IOCCC entries developed on **Centos 7**.  They would just have to
find a different platform to compile this repo to form their IOCCC compressed tarball.

**Comment by xexyl** [comment]
See my thoughts in the other thread:

This might or might not be workable. I'm afraid that it'll only be a thought for the day though - too tired to actually focus on much of anything. Basic idea is to use some Makefile magic to determine which flags to include. It might be possible.

Alternatively it could be given in documentation that under CentOS one would have to modify the Makefile slightly? That's another possibility. Since we have the makefile.local override that would be a way to go about it? 

One could actually use the Makefile feature `+=` to variables to add the additional flags.

Another thought is based on some system define checks directly give a function prototype. Maybe that's risky though but it just popped into my head since it is about implicit function declaration.

**Comment by lcn2** [comment]
> See my thoughts in the other thread:

Probably best to repeat them in this issue if they are still needed.

> This might or might not be workable. I'm afraid that it'll only be a thought for the day though - too tired to actually focus on much of anything. Basic idea is to use some Makefile magic to determine which flags to include. It might be possible.
> 
> Alternatively it could be given in documentation that under CentOS one would have to modify the Makefile slightly? That's another possibility. Since we have the makefile.local override that would be a way to go about it?
> 
> One could actually use the Makefile feature `+=` to variables to add the additional flags.
> 
> Another thought is based on some system define checks directly give a function prototype. Maybe that's risky though but it just popped into my head since it is about implicit function declaration.

The problem is that compiling with:

```sh
-std=gnu11 -D_DEFAULT_SOURCE -D_ISOC99_SOURCE -D_POSIX_C_SOURCE=200809L -D_XOPEN_SOURCE=600
```

might be a problem where the c11 C libraries are linked using headers that are processed under older standards.

At a minimum, these flags should only be enabled for **Centos 7** for those files that absolutely need those flags.

**Comment by lcn2** [comment]
If **Centos 7** does not have the `timegm(3)` function, then either someone (@xexyl) will have to write a version of that function for **Centos 7**, or consider compiling out that test when **Centos 7** is detected, or perhaps decide to abandon **Centos 7** as a platform that can compile this repo.

Again: If **Centos 7** were abandoned, then we could write a simple code test that attempts to compile with `timegm(3)`, and if that fails, tell the user that their platform is too old for the repo to support.  Such users may still submit IOCCC entries written under **Centos 7**, they would simply have to find a more recent platform on which to compile this repo in order to submit their entry.

**Comment by xexyl** [comment]
> > See my thoughts in the other thread:
> 
> Probably best to repeat them in this issue if they are still needed.
> 

They're the ones below :)

> > This might or might not be workable. I'm afraid that it'll only be a thought for the day though - too tired to actually focus on much of anything. Basic idea is to use some Makefile magic to determine which flags to include. It might be possible.
> > Alternatively it could be given in documentation that under CentOS one would have to modify the Makefile slightly? That's another possibility. Since we have the makefile.local override that would be a way to go about it?
> > One could actually use the Makefile feature `+=` to variables to add the additional flags.
> > Another thought is based on some system define checks directly give a function prototype. Maybe that's risky though but it just popped into my head since it is about implicit function declaration.
> 
> The problem is that compiling with:
> 
> ```shell
> -std=gnu11 -D_DEFAULT_SOURCE -D_ISOC99_SOURCE -D_POSIX_C_SOURCE=200809L -D_XOPEN_SOURCE=600
> ```
> 
> might be a problem where the c11 C libraries are linked using headers that are processed under older standards.
> 
> At a minimum, these flags should only be enabled for **Centos 7** for those files that absolutely need those flags.

Perhaps then some Makefile magic is in order. That or documentation update showing how to add makefile.local.

**Comment by xexyl** [comment]
> If **Centos 7** does not have the `timegm(3)` function, then either someone (@xexyl) will have to write a version of that function for **Centos 7**, or consider compiling out that test when **Centos 7** is detected, or perhaps decide to abandon **Centos 7** as a platform that can compile this repo.
> 

It does have it with specific feature test macros. But here's a cool thing. The man page might have the answer:

```man
NOTES
       The timelocal() function is equivalent to the POSIX standard function mktime(3).  There is no reason to ever use it.

       For a portable version of timegm(), set the TZ environment variable to UTC, call mktime(3) and restore the value of TZ.  Something like

           #include <time.h>
           #include <stdlib.h>

           time_t
           my_timegm(struct tm *tm)
           {
               time_t ret;
               char *tz;

               tz = getenv("TZ");
               if (tz)
                   tz = strdup(tz);
               setenv("TZ", "", 1);
               tzset();
               ret = mktime(tm);
               if (tz) {
                   setenv("TZ", tz, 1);
                   free(tz);
               } else
                   unsetenv("TZ");
               tzset();
               return ret;
           }


```

Would that work ? If so what else has to be done ? (I just want to be sure everything is completed). Also if it is sufficient will it work on other systems?

> Again: If **Centos 7** were abandoned, then we could write a simple code test that attempts to compile with `timegm(3)`, and if that fails, tell the user that their platform is too old for the repo to support. Such users may still submit IOCCC entries written under **Centos 7**, they would simply have to find a more recent platform on which to compile this repo in order to submit their entry.

That's true.

**Comment by lcn2** [comment]
Makefile magic such as:

```make
ifeq ($(target),)
target=$(shell uname -s 2>/dev/null)
endif
```
D_LEGACY=
ifeq ($(target),__WHATEVER_ uname -s_on Centos 7__)
D_LEGACY+= -D_DEFAULT_SOURCE -D_ISOC99_SOURCE -D_POSIX_C_SOURCE=200809L -D_XOPEN_SOURCE=600
endif

**Comment by xexyl** [comment]
> Makefile magic such as:
> 
> ```makefile
> ifeq ($(target),)
> target=$(shell uname -s 2>/dev/null)
> endif
> ```
> 
> D_LEGACY= ifeq ($(target),__WHATEVER_ uname -s_on Centos 7__) D_LEGACY+= -D_DEFAULT_SOURCE -D_ISOC99_SOURCE -D_POSIX_C_SOURCE=200809L -D_XOPEN_SOURCE=600 endif

Precisely.

**Comment by xexyl** [comment]
> > Makefile magic such as:
> > ```makefile
> > ifeq ($(target),)
> > target=$(shell uname -s 2>/dev/null)
> > endif
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > D_LEGACY= ifeq ($(target),__WHATEVER_ uname -s_on Centos 7__) D_LEGACY+= -D_DEFAULT_SOURCE -D_ISOC99_SOURCE -D_POSIX_C_SOURCE=200809L -D_XOPEN_SOURCE=600 endif
> 
> Precisely.

One problem: `name -s` shows just `Linux` as expected. There might be some system defined though. I can check with `cpp -dM` on some file.

**Comment by lcn2** [comment]
@xexyl: Does calls to `timegm(3)` compile under **Centos 7?**

**Comment by xexyl** [comment]
> @xexyl: Does calls to `timegm(3)` compile under **Centos 7?**

I think it would because I don't get an error on that function:

```sh
entry_util.c: In function 'timestr_eq_tstamp':
entry_util.c:1131:5: error: implicit declaration of function 'strptime' [-Werror=implicit-function-declaration]
     ptr = strptime(timestr, "%a %b %d %H:%M:%S %Y UTC", &timeptr);
     ^
entry_util.c:1131:9: error: assignment makes pointer from integer without a cast [-Werror]
     ptr = strptime(timestr, "%a %b %d %H:%M:%S %Y UTC", &timeptr);
         ^
entry_util.c: In function 'test_formed_UTC':
entry_util.c:2240:9: error: assignment makes pointer from integer without a cast [-Werror]
     ptr = strptime(str, "%a %b %d %H:%M:%S %Y UTC", &timeptr);
         ^
entry_util.c: At top level:
cc1: error: unrecognized command line option "-Wno-unreachable-code-break" [-Werror]
cc1: error: unrecognized command line option "-Wno-poison-system-directories" [-Werror]
cc1: error: unrecognized command line option "-Wno-unused-command-line-argument" [-Werror]
cc1: all warnings being treated as errors
make: *** [entry_util.o] Error 1
```

**Comment by xexyl** [comment]
Are you aware of a way to check if a function is declared and if not give it a prototype? I don't know if `#ifdef` will work on functions (macros I know it does but not sure on functions)? If it does would that be good enough? Just another thought.

**Comment by lcn2** [comment]
> > > Makefile magic such as:
> > > ```makefile
> > > ifeq ($(target),)
> > > target=$(shell uname -s 2>/dev/null)
> > > endif
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > D_LEGACY= ifeq ($(target),__WHATEVER_ uname -s_on Centos 7__) D_LEGACY+= -D_DEFAULT_SOURCE -D_ISOC99_SOURCE -D_POSIX_C_SOURCE=200809L -D_XOPEN_SOURCE=600 endif
> > 
> > 
> > Precisely.
> 
> One problem: `name -s` shows just `Linux` as expected. There might be some system defined though. I can check with `cpp -dM` on some file.

Sure.  Then something such as:

```make
ifeq ($(centos_7),)
legacy_os =$(shell command that detects Centos 7 and prints centos_7 if it does 2>/dev/null)
endif

D_LEGACY=
ifeq ($(centos_7),centos_7)
D_LEGACY+= -D_DEFAULT_SOURCE -D_ISOC99_SOURCE -D_POSIX_C_SOURCE=200809L -D_XOPEN_SOURCE=600 -DCENTOS_7
endif
```

might be needed.

**Comment by lcn2** [comment]
> Are you aware of a way to check if a function is declared and if not give it a prototype? I don't know if `#ifdef` will work on functions (macros I know it does but not sure on functions)? If it does would that be good enough? Just another thought.

We would instead suggest that CENTOS7 be a define as in:

```c
#if defined(CENTOS7)
...stuff..
#endif 
```

**Comment by xexyl** [comment]
> > Are you aware of a way to check if a function is declared and if not give it a prototype? I don't know if `#ifdef` will work on functions (macros I know it does but not sure on functions)? If it does would that be good enough? Just another thought.
> 
> We would instead suggest that CENTOS7 be a define as in:
> 
> ```c
> #if defined(CENTOS7)
> ...stuff..
> #endif 
> ```

Sure. But the problem is is there a way to do that? A quick look at `cpp -dM` didn't reveal anything to do with CentOS specifically but maybe there is a define I'm unaware of.

**Comment by lcn2** [comment]
Is a system that falls somewhat short of c11 worth supporting in this repo?

**Comment by xexyl** [comment]
> Is a system that falls somewhat short of c11 worth supporting in this repo?

Yes .. if only for solving the problem! 

But I had another thought. I'm not sure if it will be of use but we can also check the major/minor versions of the compiler. That might or might not be useful. My guess (and it's a guess **_only_**) is that it's the compiler that is having this problem.

However I'll be leaving quite soon.

**Comment by xexyl** [comment]
Interesting thought: what if autoconf can detect CentOS? I don't suggest we use it but I mean if it can then we can maybe adapt it? I got this idea because at:

https://gcc.gnu.org/onlinedocs/cpp/System-specific-Predefined-Macros.html#System-specific-Predefined-Macros

```
We are slowly phasing out all predefined macros which are outside the reserved namespace. You should never use them in new programs, and we encourage you to correct older code to use the parallel macros whenever you find it. We don’t recommend you use the system-specific macros that are in the reserved namespace, either. It is better in the long run to check specifically for features you need, using a tool such as autoconf.
```

though I think that these are actually reserved so would be okay - but it just gave me the idea.

**Comment by lcn2** [comment]
> > > Are you aware of a way to check if a function is declared and if not give it a prototype? I don't know if `#ifdef` will work on functions (macros I know it does but not sure on functions)? If it does would that be good enough? Just another thought.
> > 
> > 
> > We would instead suggest that CENTOS7 be a define as in:
> > ```c
> > #if defined(CENTOS7)
> > ...stuff..
> > #endif 
> > ```
> 
> Sure. But the problem is is there a way to do that? A quick look at `cpp -dM` didn't reveal anything to do with CentOS specifically but maybe there is a define I'm unaware of.

Well if all else fails, have a shell script (call it `legacy_os.sh` compile tiny C program that calls `timegm(3)` and `strptime(3)` and if the compile of that tiny C fails, output "legacy_os", otherwise nothing.

Then in the Makefile:

```make
ifeq ($(legacy_os),)
legacy_os=$(shell ./legacy_os.sh -s 2>/dev/null)
endif

D_LEGACY=
ifeq ($(legacy_os),legacy_os)
D_LEGACY+= -D_DEFAULT_SOURCE -D_ISOC99_SOURCE -D_POSIX_C_SOURCE=200809L -D_XOPEN_SOURCE=600 -DLEGACY_OS
endif
```

That would be a bit more general.

Then code would have to be hacked to use:

```c
#if defined(LEGACY_OS)
...
#endif
```

**Comment by xexyl** [comment]
> > > > Are you aware of a way to check if a function is declared and if not give it a prototype? I don't know if `#ifdef` will work on functions (macros I know it does but not sure on functions)? If it does would that be good enough? Just another thought.
> > > 
> > > 
> > > We would instead suggest that CENTOS7 be a define as in:
> > > ```c
> > > #if defined(CENTOS7)
> > > ...stuff..
> > > #endif 
> > > ```
> > 
> > 
> > Sure. But the problem is is there a way to do that? A quick look at `cpp -dM` didn't reveal anything to do with CentOS specifically but maybe there is a define I'm unaware of.
> 
> Well if all else fails, have a shell script (call it `legacy_os.sh` compile tiny C program that calls `timegm(3)` and `strptime(3)` and if the compile of that tiny C fails, output "legacy_os", otherwise nothing.
> 
> Then in the Makefile:
> 
> ```makefile
> ifeq ($(legacy_os),)
> legacy_os=$(shell ./legacy_os.sh -s 2>/dev/null)
> endif
> 
> D_LEGACY=
> ifeq ($(legacy_os),legacy_os)
> D_LEGACY+= -D_DEFAULT_SOURCE -D_ISOC99_SOURCE -D_POSIX_C_SOURCE=200809L -D_XOPEN_SOURCE=600 -DLEGACY_OS
> endif
> ```
> 
> That would be a bit more general.
> 
> Then code would have to be hacked to use:
> 
> ```c
> #if defined(LEGACY_OS)
> ...
> #endif
> ```

That's an idea. That could work. Hopefully tonight I can get some better sleep and I'll feel up to solving some of these problems - whether it's this issue or another issue or more than one.

Going to get some food now. Enjoy the rest of your day my friend and sleep well when you do! Make sure to activate that Easter egg today if you haven't - you'll quite appreciate the string I think!

BCNU! :)

**Comment by lcn2** [comment]
> Interesting thought: what if autoconf can detect CentOS? I don't suggest we use it but I mean if it can then we can maybe adapt it? I got this idea because at:
> 
> https://gcc.gnu.org/onlinedocs/cpp/System-specific-Predefined-Macros.html#System-specific-Predefined-Macros
> 
> ```
> We are slowly phasing out all predefined macros which are outside the reserved namespace. You should never use them in new programs, and we encourage you to correct older code to use the parallel macros whenever you find it. We don’t recommend you use the system-specific macros that are in the reserved namespace, either. It is better in the long run to check specifically for features you need, using a tool such as autoconf.
> ```
> 
> though I think that these are actually reserved so would be okay - but it just gave me the idea.

The autoconf is a world of pain that we night not wish in enter.

It would require the user to download a much of systems such as M4.

it would require the use of GPL.

**Comment by xexyl** [comment]
> > Interesting thought: what if autoconf can detect CentOS? I don't suggest we use it but I mean if it can then we can maybe adapt it? I got this idea because at:
> > https://gcc.gnu.org/onlinedocs/cpp/System-specific-Predefined-Macros.html#System-specific-Predefined-Macros
> > ```
> > We are slowly phasing out all predefined macros which are outside the reserved namespace. You should never use them in new programs, and we encourage you to correct older code to use the parallel macros whenever you find it. We don’t recommend you use the system-specific macros that are in the reserved namespace, either. It is better in the long run to check specifically for features you need, using a tool such as autoconf.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > though I think that these are actually reserved so would be okay - but it just gave me the idea.
> 
> The autoconf is a world of pain that we night not wish in enter.
> 
> It would require the user to download a much of systems such as M4.
> 
> it would require the use of GPL.
> 
> Autoconf is way overkill to solve a problem of a legacy pre-c11 system.

Agreed. I think you misunderstood. I was just thinking if we could find out _if_ it can detect CentOS then see how it does it and then maybe make use of it. That's a big if though.

**Comment by lcn2** [comment]
See commit 18cdb8a0957d458918a3b9abaff23ee6b2c55964

Added legacy_os detection code

This is a trial code in an effort to detect if a system as a problem
compiling selected functions that were added in c11.

The command:

```sh
./legacy_os.sh
```

will output nothing if all is well, or "legacy_os" is the system
has problems compiing selected functions that were added in c11.

Such code, if hosts such as Centos 7 print "legacy_os"
could be used in the Makefile as follows:

```make
ifeq ($(legacy_os),)
legacy_os=$(shell ./legacy_os.sh 2>/dev/null)
endif

D_LEGACY=
ifeq ($(legacy_os),legacy_os)
D_LEGACY+= -D_DEFAULT_SOURCE -D_ISOC99_SOURCE -D_POSIX_C_SOURCE=200809L -D_XOPEN_SOURCE=600 -DLEGACY_OS
endif
```

**Comment by xexyl** [comment]
> See commit [18cdb8a](https://github.com/ioccc-src/mkiocccentry/commit/18cdb8a0957d458918a3b9abaff23ee6b2c55964)
> 
> Added legacy_os detection code.
> 
> This is a trial code in an effort to detect if a system as a problem compiling selected functions that were added in c11.
> 
> The command:
> 
> ```shell
> ./legacy_os.sh
> ```
> 
> will output nothing if all is well, or "legacy_os" is the system has problems compiing selected functions that were added in c11.
> 
> Such code, if hosts such as Centos 7 print "legacy_os" could be used in the Makefile as follows:
> 
> ```makefile
> ifeq ($(legacy_os),)
> legacy_os=$(shell ./legacy_os.sh 2>/dev/null)
> endif
> 
> D_LEGACY=
> ifeq ($(legacy_os),legacy_os)
> D_LEGACY+= -D_DEFAULT_SOURCE -D_ISOC99_SOURCE -D_POSIX_C_SOURCE=200809L -D_XOPEN_SOURCE=600 -DLEGACY_OS
> endif
> ```
> 
> ### UPDATE 0
> @xexyl : Does `legacy_os.sh` output `legacy_os` under Centos 7?
> 
> Are there other systems, not as old, that output `legacy_os`?
> 
> Note: `legacy_os.sh` for most systems should output nothing.

Thank you for this! I will check in the morning. 

I have another system that I can check as well. I mean besides my boxes. I also have access to a rocky linux system so I can check that too. I am sure fedora is fine and I already know macOS is fine. 

I should react to this in the morning but if for some reason I don’t please ping me on this. 

Btw: I know what you mean with is this even necessary and I can even take the point as it’s older and most people wouldn’t submit entries on it. But since I made it work in the first place I would like to see that I can get it working again. Call it my child system or something like that.

**Comment by lcn2** [comment]
> I should react to this in the morning but if for some reason I don’t please ping me on this.

When you let us know if the tool prints `legacy_os` under _Centos 7_, we will make the Makefile adjustments.

**Comment by xexyl** [comment]
> > I should react to this in the morning but if for some reason I don’t please ping me on this.
> 
> When you let us know if the tool prints `legacy_os` under _Centos 7_, we will make the Makefile adjustments.

It does. However: adding to the Makefile:

```makefile
# Test for legacy OS (for example CentOS) and if necessary enable certain
# compiler flags:
ifeq ($(legacy_os),)
    legacy_os=$(shell ./legacy_os.sh 2>/dev/null)
endif

D_LEGACY=
ifeq ($(legacy_os),legacy_os)
    D_LEGACY+= -D_DEFAULT_SOURCE -D_ISOC99_SOURCE -D_POSIX_C_SOURCE=200809L -D_XOPEN_SOURCE=600 -DLEGACY_OS
endif
```

Does not work. I noticed also that `time.h` has to be included for that function which I've done in entry_util.c but this also did not work. I also noticed that `LEGACY_OS` is not used though so maybe that's why - perhaps you had something else in mind?

Given that `D_LEGACY` is added to `LEGACY_CFLAGS` which is used in the compilation rule I'm not sure why this is. I was going to work it out but I am not feeling that well today and I'll be taking a break again soon. I hope I can work on some of the other things in a bit but I hope this helps

As for other systems to test: possibly but unfortunately I forgot my password to the Rocky linux system and forgot to add an ssh key. Once that has been corrected (not sure when as he's busy) I'll test it as well.

**Comment by lcn2** [comment]
See commit 16beee56c91812b40f29248fb93a796f8cd364a3

There may be issue with other systems due to this commit, but at least it is a start.

Please report any platforms that this commit breaks on, and
feel free to suggest pull requests to fix such platforms.

```
Testing legacy / timegm detection and workarounds

Renamed legacy_os to have_timegm in this repo.

Removed ${LEGACY_CFLAGS}.  By default ${CFLAGS} is used.

By default, the -lm is not used and ${LDFLAGS} is empty.

By default, ${D_LEGACY} is empty.

By default, ${WARN_FLAGS} is -Wall -Wextra -Werror only.

Added code in Makefile to run have_timegm.sh in order to see if the
system can compile the c11 function timegm().  It if cannot, then
${TIMEGM_PROBLEM} will be set to -DTIMEGM_PROBLEM.  This is turn
will cause this to be added to ${D_LEGACY}:

    -D_DEFAULT_SOURCE -D_ISOC99_SOURCE -D_POSIX_C_SOURCE=200809L -D_XOPEN_SOURCE=600 -DTIMEGM_PROBLEM

and this is added to ${LDFLAGS}:

    -lm

and this is added to ${WARN_FLAGS):

    -Wno-unused-command-line-argument -Wno-poison-system-directories -Wno-unreachable-code-break -Wno-padded

When TIMEGM_PROBLEM is fined, entry_util.c adds a extern declaration for timegm().

Added language to the Makefile and entry_util.c explaining the
work-a-round / gross hack supporting pre-c11 systems such as Centos
7: that such systems are supported on a best guess basis and support
will go away in 2024.
```

**Comment by xexyl** [comment]
> See commit [16beee5](https://github.com/ioccc-src/mkiocccentry/commit/16beee56c91812b40f29248fb93a796f8cd364a3)
> 
> There may be issue with other systems due to this commit, but at least it is a start.
> 
> Please report any platforms that this commit breaks on, and feel free to suggest pull requests to fix such platforms.
> 
> ```
> Testing legacy / timegm detection and workarounds
> 
> Renamed legacy_os to have_timegm in this repo.
> 
> Removed ${LEGACY_CFLAGS}.  By default ${CFLAGS} is used.
> 
> By default, the -lm is not used and ${LDFLAGS} is empty.
> 
> By default, ${D_LEGACY} is empty.
> 
> By default, ${WARN_FLAGS} is -Wall -Wextra -Werror only.
> 
> Added code in Makefile to run have_timegm.sh in order to see if the
> system can compile the c11 function timegm().  It if cannot, then
> ${TIMEGM_PROBLEM} will be set to -DTIMEGM_PROBLEM.  This is turn
> will cause this to be added to ${D_LEGACY}:
> 
>     -D_DEFAULT_SOURCE -D_ISOC99_SOURCE -D_POSIX_C_SOURCE=200809L -D_XOPEN_SOURCE=600 -DTIMEGM_PROBLEM
> 
> and this is added to ${LDFLAGS}:
> 
>     -lm
> 
> and this is added to ${WARN_FLAGS):
> 
>     -Wno-unused-command-line-argument -Wno-poison-system-directories -Wno-unreachable-code-break -Wno-padded
> 
> When TIMEGM_PROBLEM is fined, entry_util.c adds a extern declaration for timegm().
> 
> Added language to the Makefile and entry_util.c explaining the
> work-a-round / gross hack supporting pre-c11 systems such as Centos
> 7: that such systems are supported on a best guess basis and support
> will go away in 2024.
> ```

I will check this out in the morning. Odds are that not having `-lm` in LDFLAGS will cause a linker error (which is why I added it in the first place) but I also see that it is added in some cases so perhaps that will be sufficient. 

If this does work under CentOS then maybe I can also clean it up a bit. Given the nature of the problem I guess this would be a kind of high priority over some of the other things?

As for other systems right now I unfortunately can only test fedora and CentOS plus macOS but I already know that will be fine since you also develop under it!

## OT: txzchk fixes and enhancements 

I am going to get ready to sleep now. Hopefully tomorrow I can actually get the txzchk fixes in. It would be nice if I can also get more test functions in but given how I have been feeling lately this might be a bit optimistic. Depends on how I sleep too. 

## OT: rot13ss - my private tool

I have to do some thinking out on the private tool for foo but hopefully I can get that in a good enough state to share as well. To be clear it could be shared now with some explanation of how to use it for multiple strings (single strings works fine) but it’s also quite messy (more messy from the changes of today). The primary thing that has to be fixed is multiple strings especially undoing it. The original transliteration is mostly okay and certainly works for getting the right strings (just if you use more than one string you have to slightly modify the output. 

The reversal is more problematic because it has to distinguish `\\"` (or `\"`) from just `"` which will take some kind of magic which really takes enough focus (and energy and motivation) to sort out. If I scp it to the linux machines I could use gdb and that would probably help my exhausted state (it’s amazing how being tired can dramatically change the ability to work out problems!) - I don’t really know anything about macOS debugging (lldb?). I probably should learn about it one of these days. 

Of course if I were to share more of it with you you might have some ideas on how to deal with this situation. Perhaps you have some ideas just with that explanation? If you do I would love to hear them! (I guess I could go the disgusting route and use my newfound knowledge of flex and bison but I don’t want to make myself even more insane 😀 and I prefer to do it manually even if it means I will sometimes take longer. 

Of course it is also possible that I don’t bother with the reversal but the idea of being able to automagically generate the array sounds quite useful. Any thoughts on that also welcome!

## OT: chkentry (last topic for now!)

Hopefully I can get more done with it sometime soon too. I could swear I already wrote this but I don’t see it above. Very possibly I wrote it in the other thread right before I shut down the laptop: not sure. But I am sure I wrote something about it so I won’t repeat it here. 

And with that I am going to try sleeping. I hope you had a great sleep (I guess you’re already sleeping) and I hope you have a great day when you wake up. 

Cheers and until next time!

**Comment by xexyl** [comment]
Oh one more thing real quick. CentOS actually does support c11 which makes this all the more strange. Maybe it’s gnu11 ? But I am pretty sure it is okay. I should check both clang AND gcc. Only just thought of that. Perhaps that could simplify the problem? 

But even so the point of CentOS is stability for servers and it indeed is a server. So it’s expected that some things might need feature test macros etc. 

Anyway hopefully I can do something with the commit but for now I am going to sleep. If you have any thoughts on any of this (including and maybe especially rot13ss - the undo mode and whether or not I should bother with making a generator [0]) I more than welcome them. 

[0] But of course I might still want to do it simply because I thought of it and it would make it easier for both of us. Well we’ll see. 

Going now. More from me tomorrow in some form or another (most likely several forms)!

**Comment by xexyl** [comment]
Here's a strange thing. The linking works but what does not is this:

```sh
cc -std=gnu11 -O0 -g -pedantic  -D_DEFAULT_SOURCE -D_ISOC99_SOURCE -D_POSIX_C_SOURCE=200809L -D_XOPEN_SOURCE=600 -DTIMEGM_PROBLEM -Wall -Wextra -Werror -Wno-unused-command-line-argument -Wno-poison-system-directories -Wno-unreachable-code-break -Wno-padded  -lm chkentry.c -c
In file included from chkentry.h:44:0,
                 from chkentry.c:35:
foo.h:47:12: error: 'foo' defined but not used [-Werror=unused-variable]
 static int foo[] = { 'B', 'E', 'Q', 'V', 'D', 'U',
            ^
cc1: error: unrecognized command line option "-Wno-unreachable-code-break" [-Werror]
cc1: error: unrecognized command line option "-Wno-poison-system-directories" [-Werror]
cc1: error: unrecognized command line option "-Wno-unused-command-line-argument" [-Werror]
cc1: all warnings being treated as errors
make: *** [chkentry.o] Error 1
```

Removing static makes

```sh
foo.o:(.data+0x0): multiple definition of `foo'
chkentry.o:(.data+0x0): first defined here
```

And that's because in `chkentry.h`:

```c
/*
 * foo - bar
 */
#include "foo.h"
```

...which it needs because of the function it calls. However it does not need the `foo` array. So I can think of a number of ways to solve this but since one is something you were not in favour of before I'll list them here.

- Move the foo array into foo.c. I'm against this because I want the array out of sight.
- Have an include guard in foo.h so that if `FOO_C` is defined it will declare the foo array; else it won't be declared. You don't like this way so we can say maybe not to do this one.
- Add the prototype of the function to chkentry.h and remove from it inclusion of foo.h with a note saying something (but what?). Maybe there doesn't need to be a note? I'll do that for now and if you want a note you can add one as it works and since the function is used in chkentry.c it makes sense it could belong there even though it's defined in foo.c.

Doing this the code compiles under CentOS. I'm afraid that I'm too tired to do much of anything today though. I slept in way late today but I got to sleep later. I'm not even sure if I can get the txzchk changes in today. I feel so exhausted; I feel like I've been drugged or something: very fatigued.

But I'll prepare a pull request soon so that the code compiles under CentOS.

**Comment by xexyl** [comment]
See https://github.com/ioccc-src/mkiocccentry/pull/313/commits/47e80f032abbcb25dc82433d1e18198ae44529f6 for a fix. Not tested this fix in fedora but I suspect it'll work fine (it does work under macOS of course).

**Comment by lcn2** [comment]
> Have an include guard in foo.h so that if FOO_C is defined it will declare the foo array; else it won't be declared. You don't like this way so we can say maybe not to do this one.

This might be the best of the 3 workarounds you offered. It is small, compact and well the file is _foo_ so why not _fubar_ things a bit further? 😄

**Comment by xexyl** [comment]
> > Have an include guard in foo.h so that if FOO_C is defined it will declare the foo array; else it won't be declared. You don't like this way so we can say maybe not to do this one.
> 
> This might be the best of the 3 workarounds you offered. It is small, compact and well the file is _foo_ so why not _fubar_ things a bit further? 😄

Well that's something to consider too. I only went the way I did from your previous disliking of that idea. Do you want to do that then?

**Comment by xexyl** [comment]
I'll do that now as I haven't pulled the changes on the CentOS system. Then doing something else. Okay well I might have some more quotes to add but nothing productive anyway.

**Comment by lcn2** [comment]
The recent commit 985b4f2c9f4ae1ece3c1504e64bf9845f60c417d (Merge pull request #313 from xexyl/updates) is fine for macOS.

**Comment by xexyl** [comment]
> The recent commit [985b4f2](https://github.com/ioccc-src/mkiocccentry/commit/985b4f2c9f4ae1ece3c1504e64bf9845f60c417d) (Merge pull request #313 from xexyl/updates) is fine for macOS.

I did the pull request from it so I should hope so :-)

**Comment by xexyl** [comment]
I just committed a fix with something possibly funny but I am going to add some quotes as well soon .. then will do a pull request.

**Comment by xexyl** [comment]
Okay I admit it: It is funny but only those with my tool will be able to understand it. :-)

**Comment by xexyl** [comment]
Are we good to close this issue? Or do you want me to see if I can clean it up some (when I have more energy that is)?

As for the tool: I have had some ideas that might help. Not sure if I have the energy now to do it but I'm hoping to later today if I have time before my doctor appointment. I right now would like to work on the chkentry functions but I'm not sure if I am yet awake enough for that (which might mean it's better to work on the tool code now: not sure yet .. but I do know I will try sleeping again in a bit: once the backup is done and the drive has cooled down a bit to put away).

**Comment by lcn2** [comment]
> Are we good to close this issue? 

It appears we are "code complete" for this issue. 

Perhaps some of the language about the issue, about CenTOS 7, and/or c11 might be improved? We recall you might have suggested a better was to express the issue. 

If that is true, feel free to submit a pull request. Otherwise, if no text changes are needed, please so advise so we can close this issue.

**Comment by xexyl** [comment]
> > Are we good to close this issue?
> 
> It appears we are "code complete" for this issue.
> 
> Perhaps some of the language about the issue, about CenTOS 7, and/or c11 might be improved? We recall you might have suggested a better was to express the issue.
> 
> If that is true, feel free to submit a pull request. Otherwise, if no text changes are needed, please so advise so we can close this issue.

Well here's my thinking right now:

I am quite tired and have a busy day but since before the repo is to be used more widely I am going to go over everything to fix typos etc. this can be done later. Thus I don't see why you have to have this open. If necessary I can always ask you about certain things later on?

**Comment by xexyl** [comment]
Well some changes broke this fix. I'm not awake enough to look at it and maybe it's lower priority (not sure) but:

```sh
cc -std=gnu11 -O0 -g -pedantic  -Wall -Wextra -Werror  entry_util.c -c
entry_util.c: In function 'timestr_eq_tstamp':
entry_util.c:1618:5: error: implicit declaration of function 'strptime' [-Werror=implicit-function-declaration]
     ptr = strptime(timestr, "%a %b %d %H:%M:%S %Y UTC", &timeptr);
     ^
entry_util.c:1618:9: error: assignment makes pointer from integer without a cast [-Werror]
     ptr = strptime(timestr, "%a %b %d %H:%M:%S %Y UTC", &timeptr);
         ^
entry_util.c: In function 'test_formed_UTC':
entry_util.c:2840:9: error: assignment makes pointer from integer without a cast [-Werror]
     ptr = strptime(str, "%a %b %d %H:%M:%S %Y UTC", &timeptr);
         ^
cc1: all warnings being treated as errors
make: *** [entry_util.o] Error 1
```

Just so you know ... Maybe I can look at it later but we'll see. I hope to go back to sleep in a bit but most likely it'll only be rest. Still it was a much better night last night (not that it was good though).

Hope you're having a nice day!

**Comment by xexyl** [comment]
... fixed it, will make a commit soon.

**Comment by xexyl** [comment]
...on second look, I forgot I had done a minor change to the Makefile which would mean it would not work on all hosts. Even so I have some ideas.

**Comment by xexyl** [comment]
## UPDATE 3 (Possible fix)

It gets uglier but a possible fix. it seems that whereas ``` works for macOS and I'm sure linux it's not working properly (in this case) in CentOS - but `shell` does. Is it permissible to have for this part of the Makefile it use `shell` as it does seem to work under macOS too ? If so this would solve the problem.

**Comment by lcn2** [comment]
> ## UPDATE 3 (Possible fix)
> It gets uglier but a possible fix. it seems that whereas backticks work for macOS and I'm sure linux it's not working properly (in this case) in CentOS - but `shell` does. 

It is time to abandon support of CentOS 7?

> Is it permissible to have for this part of the Makefile it use `shell` as it does seem to work under macOS too ? If so this would solve the problem.

What part of the Makefile?  Perhaps you can post a patch as a comment?

**Comment by xexyl** [comment]
> > ## UPDATE 3 (Possible fix)
> > It gets uglier but a possible fix. it seems that whereas backticks work for macOS and I'm sure linux it's not working properly (in this case) in CentOS - but `shell` does.
> 
> It is time to abandon support of CentOS 7? It night be but lets give it one more try ...
> 
> > Is it permissible to have for this part of the Makefile it use `shell` as it does seem to work under macOS too ? If so this would solve the problem.
> 
> What part of the Makefile? Perhaps you can post a patch as a comment?

```diff
-D_LEGACY=
-ifeq ($(TIMEGM_PROBLEM),)
-TIMEGM_PROBLEM="`./have_timegm.sh 2>/dev/null`"
-endif
-
+D_LEGACY :=
+HAVE_TIMEGM=$(shell ./have_timegm.sh 2>/dev/null)
 
```

The important part being using shell in that one place. You can probably ignore the rest as that was before I discovered that the expansion being incorrect (I just had another thought that might work if this is not allowed here though I'm not certain). But I think it's okay to use `shell` instead of the backticks in this place?

**Comment by xexyl** [comment]
> See commit [005bfe8](https://github.com/ioccc-src/mkiocccentry/commit/005bfe83b404cf995ba35fa9ef36f084cc38d63b)
> 
> Perhaps this will allow CentOS 7 to be supported a bit longer? :-)

Indeed it works.

**Comment by xexyl** [comment]
... leaving for the day. More from me tomorrow in some form or another I'm sure.
